File: WorldFeed/Server/src/Services/Post/WorldFeed.Post.API/WebLogic/TwitterClientHandler.cs
namespace WorldFeed.Post.API.WebLogic
{
    using System;
    using System.Net.Http;
    using System.Net.Http.Headers;
    using System.Threading;
    using System.Threading.Tasks;

    using WorldFeed.Common.Extensions;
    using WorldFeed.Common.Public.Models.Interfaces;
    using WorldFeed.Post.API.Application.Query;

    public class TwitterClientHandler : HttpClientHandler, ITwitterClientHandler
    {
        private readonly Action<ITwitterQuery, HttpRequestMessage> action;
        private readonly Func<ITwitterQuery, HttpRequestMessage, string> func;

        // protected IOAuthWebRequestGenerator WebRequestGenerator { get; }

        private ITwitterQuery twitterQuery;

        public TwitterClientHandler()  // IOAuthWebRequestGenerator oAuthWebRequestGenerator
        {
            UseCookies = false;
            UseDefaultCredentials = false;

            // WebRequestGenerator = oAuthWebRequestGenerator;
        }

        public TwitterClientHandler(Action<ITwitterQuery, HttpRequestMessage> action) : this()
        {
            this.action = action;
        }

        public TwitterClientHandler(Func<ITwitterQuery, HttpRequestMessage, string> func) : this()
        {
            this.func = func;
        }

        public ITwitterQuery TwitterQuery
        {
            get => this.twitterQuery;
            set
            {
                this.twitterQuery = value;

                if (value != null)
                {
                    Proxy = value.ProxyConfig;

                    if (Proxy != null)
                    {
                        UseProxy = true;
                    }
                }
                else
                {
                    Proxy = null;
                    UseProxy = false;
                }
            }
        }

        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            return SendAsync(this.twitterQuery, request, cancellationToken);
        }

        protected virtual async Task<HttpResponseMessage> SendAsync(ITwitterQuery twitterQuery, HttpRequestMessage request, CancellationToken cancellationToken)
        {
            this.action?.Invoke(twitterQuery, request);

            if (twitterQuery.AuthorizationHeader == null)
            {
                if (this.func != null)
                {
                    twitterQuery.AuthorizationHeader = this.func(twitterQuery, request);
                }
                else
                {
                    // await WebRequestGenerator.SetTwitterQueryAuthorizationHeaderAsync(twitterQuery).ConfigureAwait(false);
                }
            }

            return await SendAsync(request, cancellationToken, twitterQuery.AuthorizationHeader).ConfigureAwait(false);
        }

        protected Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken, string authorizationHeader)
        {
            request.Headers.Add("User-Agent", "Tweetinvi/5.0.0-alpha-6");
            request.Headers.ExpectContinue = false;
            request.Headers.CacheControl = new CacheControlHeaderValue { NoCache = true };
            request.Headers.Add("Authorization", authorizationHeader);
            request.Version = new Version("1.1");

            this.twitterQuery?.AcceptHeaders.ForEach(contentType =>
            {
                request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue(contentType));
            });

            this.twitterQuery?.CustomHeaders.ForEach(customHeader =>
            {
                if (customHeader.Behaviour == CustomHeaderWill.RemoveGeneratedHeaders)
                {
                    request.Headers.Remove(customHeader.Key);
                    return;
                }

                if (customHeader.Behaviour == CustomHeaderWill.OverrideGeneratedHeaders)
                {
                    if (request.Headers.Contains(customHeader.Key))
                    {
                        request.Headers.Remove(customHeader.Key);
                    }
                }

                request.Headers.TryAddWithoutValidation(customHeader.Key, customHeader.Values);
            });

            return base.SendAsync(request, cancellationToken);
        }
    }
}

File: Sharebook/Client/src/app/webLogic/TwitterClientHandler.ts
import {HttpEvent, HttpHandler, HttpRequest, HttpResponse} from "@angular/common/http";

import {ITwitterClientHandler} from "../core/Core/Web/ITwitterClientHandler";
import {IOAuthWebRequestGenerator} from "../core/Core/Web/IOAuthWebRequestGenerator";
import {ITwitterQuery} from '../core/Public/Models/Interfaces/ITwitterQuery';
import {CustomHeaderWill} from "../core/Public/Models/Interfaces/CustomRequestHeaders";
import {Observable} from "rxjs";
import Type from "typescript-dotnet-commonjs/System/Types";

export class TwitterClientHandler extends HttpHandler implements ITwitterClientHandler {
  private readonly _action: (twitterQuery: ITwitterQuery, httpRequestMessage: HttpRequest<any>) => void;
  private readonly _func: (twitterQuery: ITwitterQuery, httpRequestMessage: HttpRequest<any>) => string;
  protected WebRequestGenerator: IOAuthWebRequestGenerator;

  private _twitterQuery: ITwitterQuery;

  constructor(oAuthWebRequestGenerator: IOAuthWebRequestGenerator, func?: (twitterQuery: ITwitterQuery, httpRequestMessage: HttpRequest<any>) => string | void,
              isFuncReturningVoid?: boolean) {
    super();

    if (func) {
      if (isFuncReturningVoid) {
        this._action = func;
      } else {
        this._func = func as ((twitterQuery: ITwitterQuery, httpRequestMessage: HttpRequest<HttpRequest<any>>) => string);
      }
    }

    // super.useCookies = false;
    // super.useDefaultCredentials = false;

    this.WebRequestGenerator = oAuthWebRequestGenerator;
  }

  get twitterQuery(): ITwitterQuery {
    return this._twitterQuery;
  }

  set twitterQuery(value: ITwitterQuery) {
    this._twitterQuery = value;

    // if (value != null) {
    //   super.proxy = value.proxyConfig;
    //
    //   if (Proxy != null) {
    //     UseProxy = true;
    //   }
    // }
    // else {
    //   super.Proxy = null;
    //   super.UseProxy = false;
    // }
  }

  protected async SendAsync(request: HttpRequest<any>, twitterQueryOrAuthorizationHeader?: ITwitterQuery | string): Promise<HttpResponse<any>> {
    let twitterQueryCurrent: ITwitterQuery;
    let authorizationHeaderCurrent: string;

    if (!twitterQueryOrAuthorizationHeader) {
      twitterQueryCurrent = this._twitterQuery;
    }

    if (Type.isString(twitterQueryOrAuthorizationHeader)) {
      authorizationHeaderCurrent = twitterQueryOrAuthorizationHeader;
      twitterQueryCurrent = this._twitterQuery;
    } else {
      twitterQueryCurrent = twitterQueryOrAuthorizationHeader;

      this._action(twitterQueryCurrent, request);

      if (this._func != null) {
        twitterQueryCurrent.authorizationHeader = this._func(twitterQueryCurrent, request);
      } else {
        await this.WebRequestGenerator.setTwitterQueryAuthorizationHeaderAsync(twitterQueryCurrent);
      }

      authorizationHeaderCurrent = twitterQueryCurrent.authorizationHeader;
    }

    request.headers.set("User-Agent", "Tweetinvi/5.0.0-beta-1");
    // request.headers.ExpectContinue = false;
    // request.headers.CacheControl = new CacheControlHeaderValue { NoCache = true };
    request.headers.set("Authorization", authorizationHeaderCurrent);
    // request.version = new Version("1.1");

    twitterQueryCurrent?.acceptHeaders.forEach(contentType => {
      // request.headers.Accept.Add(new MediaTypeWithQualityHeaderValue(contentType));
    });

    [...this._twitterQuery?.customHeaders].forEach(customHeader => {
      if (customHeader.behaviour === CustomHeaderWill.RemoveGeneratedHeaders) {
        request.headers.delete(customHeader.key);
        return;
      }

      if (customHeader.behaviour === CustomHeaderWill.OverrideGeneratedHeaders) {
        if (request.headers.has(customHeader.key)) {
          request.headers.delete(customHeader.key);
        }
      }

      request.headers.set(customHeader.key, customHeader.values); // TryAddWithoutValidation
    });

    // return base.SendAsync(request);
    return null;
  }

  handle(req: HttpRequest<any>): Observable<HttpEvent<any>> {
    return undefined;
  }
}
